### PR TITLE
[Merged by Bors] - feat: add Prod.associated_iff

### PR DESCRIPTION
**Pull request by EmilieUthaiwat**
---

Given two monoids M and N, two elements of the product of M and N are associated if and only if their projections are associated. 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Associated.lean [hunk 1]
```diff
@@ -530,6 +531,13 @@ theorem Associated.pow_pow [CommMonoid α] {a b : α} {n : ℕ} (h : a ~ᵤ b) :
   convert h.mul_mul ih <;> rw [pow_succ]
 #align associated.pow_pow Associated.pow_pow
 
+theorem Prod.associated_iff {M N : Type*} [Monoid M] [Monoid N] {x z : M × N} :
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Associated.lean:534]: You can use `~ᵤ` instead of `Associated`.

Comment (EmilieUthaiwat) [reply Mathlib/Algebra/Associated.lean:534]: I have made the modifications. Thank you!

**Comment by EmilieUthaiwat** [comment]
I thought that it was reasonable to add Prod.associated_iff to the file Mathlib.Algebra.Associated (as it lists several cases where elements are associated) but if Mathlib.Algebra.Group.Prod should not be imported to the latter, I can put Prod.associated_iff in the file Mathlib.Algebra.BigOperators.Associated instead.

File: Mathlib/Algebra/Associated.lean [hunk 2]
```diff
@@ -530,6 +531,13 @@ theorem Associated.pow_pow [CommMonoid α] {a b : α} {n : ℕ} (h : a ~ᵤ b) :
...
+    x ~ᵤ z ↔ x.1 ~ᵤ z.1 ∧ x.2 ~ᵤ z.2 :=
+  ⟨fun ⟨u, hu⟩ => ⟨⟨(MulEquiv.prodUnits.toFun u).1, (Prod.eq_iff_fst_eq_snd_eq.1 hu).1⟩,
+    ⟨(MulEquiv.prodUnits.toFun u).2, (Prod.eq_iff_fst_eq_snd_eq.1 hu).2⟩⟩,
+      fun ⟨⟨u₁, h₁⟩, ⟨u₂, h₂⟩⟩ =>
+        ⟨MulEquiv.prodUnits.invFun (u₁, u₂), Prod.eq_iff_fst_eq_snd_eq.2 ⟨h₁, h₂⟩⟩⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:539]: I think this should deindent a bit:
```suggestion
  fun ⟨⟨u₁, h₁⟩, ⟨u₂, h₂⟩⟩ =>
    ⟨MulEquiv.prodUnits.invFun (u₁, u₂), Prod.eq_iff_fst_eq_snd_eq.2 ⟨h₁, h₂⟩⟩⟩
```

File: Mathlib/Algebra/Associated.lean [hunk 3]
```diff
@@ -6,6 +6,7 @@ Authors: Johannes Hölzl, Jens Wagemaker
 import Mathlib.Algebra.Divisibility.Basic
 import Mathlib.Algebra.GroupPower.Lemmas
 import Mathlib.Algebra.Parity
+import Mathlib.Algebra.Group.Prod
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:9]: This is a pretty low file in the hierarchy; I'm not sure if we want to add more imports to it. (See also #7200)

Comment (EmilieUthaiwat) [reply Mathlib/Algebra/Associated.lean]: Well noted. I have moved the result to `Mathlib.Algebra.BigOperators.Associated`.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

File: Mathlib/Algebra/BigOperators/Associated.lean [hunk 4]
```diff
@@ -50,6 +50,13 @@ theorem exists_mem_finset_dvd {s : Finset β} {f : β → α} : p ∣ s.prod f 
 
 end Prime
 
+theorem Prod.associated_iff {M N : Type*} [Monoid M] [Monoid N] {x z : M × N} :
+    x ~ᵤ z ↔ x.1 ~ᵤ z.1 ∧ x.2 ~ᵤ z.2 :=
+  ⟨fun ⟨u, hu⟩ => ⟨⟨(MulEquiv.prodUnits.toFun u).1, (Prod.eq_iff_fst_eq_snd_eq.1 hu).1⟩,
+    ⟨(MulEquiv.prodUnits.toFun u).2, (Prod.eq_iff_fst_eq_snd_eq.1 hu).2⟩⟩,
+  fun ⟨⟨u₁, h₁⟩, ⟨u₂, h₂⟩⟩ =>
+    ⟨MulEquiv.prodUnits.invFun (u₁, u₂), Prod.eq_iff_fst_eq_snd_eq.2 ⟨h₁, h₂⟩⟩⟩
+
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Associated.lean:59]: This seems like a very strange file for this lemma, as it's not about a big operator!

Comment (EmilieUthaiwat) [reply Mathlib/Algebra/BigOperators/Associated.lean:59]: The whole story is that I need a file which talks about associated elements and which imports both [Mathlib.Algebra.Associated](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Associated.html) and [Mathlib.Algebra.BigOperators.Associated](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/BigOperators/Associated.html). Do you think I should move the result somewhere else?

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Associated.lean:59]: Why do you need a file that imports Mathlib.Algebra.BigOperators.Associated?

Comment (EmilieUthaiwat) [reply Mathlib/Algebra/BigOperators/Associated.lean:59]: I copied and pasted the wrong link. I am sorry. I need a file which imports both [Mathlib.Algebra.Associated](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Associated.html) and [Mathlib.Algebra.Group.Prod](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Prod.html).

**Pull request closed without merging**